Userland/Libraries/LibSoftGPU/Device.cpp

```diff
@@ -834,29 +856,43 @@ void Device::submit_triangle(const Triangle& triangle, Vector<size_t> const& ena
 
         // Calculate fog
         // Math from here: https://opengl-notes.readthedocs.io/en/latest/topics/texturing/aliasing.html
+
+        // FIXME: exponential fog is not vectorized, we should add a SIMD exp function that calculates an approximation.
         if (m_options.fog_enabled) {
-            float factor = 0.0f;
+            auto factor = expand4(0.0f);
             switch (m_options.fog_mode) {
             case FogMode::Linear:
-                factor = (m_options.fog_end - fog_depth) / (m_options.fog_end - m_options.fog_start);
-                break;
-            case FogMode::Exp:
-                factor = expf(-m_options.fog_density * fog_depth);
-                break;
-            case FogMode::Exp2:
-                factor = expf(-((m_options.fog_density * fog_depth) * (m_options.fog_density * fog_depth)));
+                factor = (m_options.fog_end - quad.fog_depth) / (m_options.fog_end - m_options.fog_start);
                 break;
+            case FogMode::Exp: {
+                auto argument = -m_options.fog_density * quad.fog_depth;
+                factor = f32x4 {
+                    expf(argument[0]),
+                    expf(argument[1]),
+                    expf(argument[2]),
+                    expf(argument[3]),
```

Comment:
maybe make a helper for that, in case we make a proper implementation for it
(will be worth it after 100 iteration or so, which we likely have)

Reply:
Good point, that will probably go into SIMDExtras.h then or probably even SIMDMath.h. I'll go over all commits anyway to clean them up. If you find anything else just drop a comment.

Reply:
Added AK/SIMDMath.h